src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,673 @@
+							   "initialized, please retry")));
+	}
+
+	Assert(LogicalClockShmem->clockInitialized == CLOCKSTATE_UNINITIALIZED);
+
+	/*
+	 * Set the flag before initialization to prevent multiple backends
+	 * attempting the same.
+	 */
+	LogicalClockShmem->clockInitialized = CLOCKSTATE_INIT_INPROGRESS;
+
+	LWLockRelease(&LogicalClockShmem->clockLock);
+
+	PG_TRY();
+	{
+		/* Initialize sequences for the first time */
+		text *logicalSequenceName = cstring_to_text(CLUSTER_CLOCK_LOGICAL_SEQUENCE_NAME);
+
+		/*
+		 * Cache the Oids in global variables, no spinlock is needed as only one
+		 * backend should be in this path at any time
+		 */
+		LogicalClockShmem->logicalSequenceOid =
+			ResolveRelationId(logicalSequenceName, false);
+
+		if (LogicalClockShmem->logicalSequenceOid == InvalidOid)
+		{
+			ereport(ERROR, (errmsg("Clock related sequence is missing")));
+		}
+
+		/* Start with the wall clock value */
+		ClusterClock *epochValue = GetEpochTimeAsClock();
+
+		LogicalClockShmem->clusterClockValue = *epochValue;
+
+		/*
+		 * Retrieve the highest clock value persisted in the sequence.
+		 */
+		ClusterClock persistedMaxClock = { 0 };
+
+		/* We will get one more than the persisted value, but that's harmless */
+		persistedMaxClock.logical =
+			DirectFunctionCall1(nextval_oid,
+								ObjectIdGetDatum(LogicalClockShmem->logicalSequenceOid));
+
+		/*
+		 * Sequence 1 indicates no prior clock timestamps on this server, retain
+		 * the wall clock i.e. no adjustment needed.
+		 */
+		if (persistedMaxClock.logical != 1)
+		{
+			ereport(NOTICE, (errmsg("adjusting the clock with persisted value: "
+									"<logical(%lu) and counter(%u)>",
+									persistedMaxClock.logical,
+									persistedMaxClock.counter)));
+
+			/*
+			 * As sequences are not transactional, this will protect us from crashes after the logical
+			 * increment and before the counter increment.
+			 */
+			persistedMaxClock.logical++;
+			persistedMaxClock.counter = 0;
+
+			/*
+			 * Adjust the local clock according to the most recent
+			 * clock stamp value persisted in the catalog.
+			 */
+			AdjustLocalClock(&persistedMaxClock);
+		}
+	}
+	PG_CATCH();
+	{
+		LWLockAcquire(&LogicalClockShmem->clockLock, LW_EXCLUSIVE);
+
+		LogicalClockShmem->clockInitialized = CLOCKSTATE_UNINITIALIZED;
+		memset(&LogicalClockShmem->clusterClockValue, 0, sizeof(ClusterClock));
+
+		LWLockRelease(&LogicalClockShmem->clockLock);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	LWLockAcquire(&LogicalClockShmem->clockLock, LW_EXCLUSIVE);
+
+	LogicalClockShmem->clockInitialized = CLOCKSTATE_INITIALIZED;
+
+	LWLockRelease(&LogicalClockShmem->clockLock);
+}
+
+
+/*
+ * citus_get_cluster_clock() is an UDF that returns a monotonically increasing
+ * logical clock. Clock guarantees to never go back in value after restarts, and
+ * makes best attempt to keep the value close to unix epoch time in milliseconds.
+ */
+Datum
+citus_get_cluster_clock(PG_FUNCTION_ARGS)
```

Comment:
I'm not good at naming, but I think I'd probably use/save this name for the follow up UDF we are going to have. This one could probably be called `citus_get_node_clock ` or such